File: product-microservice/api/product_purchase_api.go
package api

import (
	"product/dto"
	"product/mapper"
	"product/middleware"
	"product/service"
	"strconv"

	"net/http"

	"github.com/gin-gonic/gin"
)


type ProductPurchaseAPI struct {
	IProductPurchaseService service.IProductPurchaseService
}

func NewProductPurchaseAPI(productPurchaseService service.IProductPurchaseService) ProductPurchaseAPI {
	return ProductPurchaseAPI{IProductPurchaseService: productPurchaseService}
}

func (productPurchaseApi *ProductPurchaseAPI) GetPurchaseHistory(c *gin.Context) {
	userId, err := strconv.Atoi(c.Query("userId"))
	page, err := strconv.Atoi(c.Query("page"))
	pageSize, err := strconv.Atoi(c.Query("pageSize"))
	if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
	}

	productPurchases := productPurchaseApi.IProductPurchaseService.GetPurchaseHistory(userId, page, pageSize)
	c.JSON(http.StatusOK, mapper.ToProductPurchaseDTOs(productPurchases))
}

func (productPurchaseApi *ProductPurchaseAPI) GetNumberOfRecordsPurchaseHistory(c *gin.Context) {
	userId, err := strconv.Atoi(c.Query("userId"))
	if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
	}
	numberOfRecords := productPurchaseApi.IProductPurchaseService.GetNumberOfRecordsPurchaseHistory(userId)
	c.JSON(http.StatusOK, numberOfRecords)
}

func (productPurchaseApi *ProductPurchaseAPI) CheckIfProductIsPaidFor(c *gin.Context) {
	productId, err := strconv.Atoi(c.Query("productId"))
	if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
	}

	userData := middleware.GetUserData(c)
	isPaidFor := productPurchaseApi.IProductPurchaseService.CheckIfProductIsPaidFor(productId, userData.Id)
	c.JSON(http.StatusOK, isPaidFor)
}

func (productPurchaseApi *ProductPurchaseAPI) ConfirmPurchase(c *gin.Context) {
	var productPurchaseDTO dto.ProductPurchaseDTO
	err := c.BindJSON(&productPurchaseDTO)
	if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
	}

	userData := middleware.GetUserData(c)
	error := productPurchaseApi.IProductPurchaseService.ConfirmPurchase(productPurchaseDTO, userData.Id)

	if error == nil {
		c.JSON(http.StatusOK, "Purchase successful")
	} else  {
		c.JSON(http.StatusBadRequest, error.Error())
	}
}

func (productPurchaseApi *ProductPurchaseAPI) SendPurchaseConfirmationMail(c *gin.Context) {
	var productPurchaseDTO dto.ProductPurchaseDTO
	err := c.BindJSON(&productPurchaseDTO)
	if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
	}

	userData := middleware.GetUserData(c)
	msg := productPurchaseApi.IProductPurchaseService.SendPurchaseConfirmationMail(productPurchaseDTO, userData.Id)
	if msg == "" {
		c.Status(200)
	} else  {
		c.JSON(http.StatusBadRequest, msg)
	}
}


func (productPurchaseApi *ProductPurchaseAPI) ConfirmPayment(c *gin.Context) {
	var productPurchaseDTO dto.ProductPurchaseDTO
	err := c.BindJSON(&productPurchaseDTO)
	if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
	}

	msg := productPurchaseApi.IProductPurchaseService.ConfirmPayment(productPurchaseDTO)
	if msg == "" {
		c.JSON(http.StatusOK, "Purchase is paid")
	} else  {
		c.JSON(http.StatusBadRequest, msg)
	}
}

func (productPurchaseApi *ProductPurchaseAPI) SendPurchasedDigitalVideoGames(c *gin.Context) {
	var productPurchaseDTO dto.ProductPurchaseDTO
	err := c.BindJSON(&productPurchaseDTO)
	if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
	}

	msg := productPurchaseApi.IProductPurchaseService.SendPurchasedDigitalVideoGames(productPurchaseDTO)
	if msg == "" {
		c.Status(200)
	} else  {
		c.JSON(http.StatusBadRequest, msg)
	}
}

func (productPurchaseApi *ProductPurchaseAPI) GetProductAlertByProductIdAndUserId(c *gin.Context) {
	productId, err := strconv.Atoi(c.Query("productId"))
	if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
	}
	userData := middleware.GetUserData(c)
	productAlert, err := productPurchaseApi.IProductPurchaseService.GetProductAlertByProductIdAndUserId(userData.Id, productId)

	if err == nil {
		c.JSON(http.StatusOK, productAlert)
	} else {
		c.JSON(http.StatusBadRequest, err)
	}
}

func (productPurchaseApi *ProductPurchaseAPI) AddProductAlert(c *gin.Context) {
	productId, err := strconv.Atoi(c.Query("productId"))
	if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
	}

	userData := middleware.GetUserData(c)
	msg := productPurchaseApi.IProductPurchaseService.AddProductAlert(userData.Id, productId);

	if msg == "" {
		c.JSON(http.StatusOK, "You will be notified via email when product is in stock.")
	} else {
		c.JSON(http.StatusBadRequest, msg)
	} 
}

func (productPurchaseApi *ProductPurchaseAPI) NotifyProductAvailability(c *gin.Context) {
	productId, err := strconv.Atoi(c.Query("productId"))
	if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
	}

	resp, err := productPurchaseApi.IProductPurchaseService.NotifyProductAvailability(productId);

	if err == nil {
		c.JSON(http.StatusOK, resp)
	} else {
		c.JSON(http.StatusBadRequest, err)
	}
}

File: api-gateway/api/product_microservice/product_purchase_api.py
import requests
from flask import request, jsonify
from utils import token_utils, role
from utils.routes.product_microservice import product_purchase_api_routes


@token_utils.authorization_required(roles=[role.ROLE_USER])
def check_if_product_is_paid_for():
    args = request.args.to_dict()
    product_id = args.get("productId")
    headers = request.headers
    r = requests.get(product_purchase_api_routes.BASE + product_purchase_api_routes.API +
                        product_purchase_api_routes.CHECK_IF_PRODUCT_IS_PAID_FOR +
                        "?productId={}".format(product_id), headers=headers)
    resp = jsonify(r.json())
    resp.status_code = r.status_code
    return resp


@token_utils.authorization_required(roles=[role.ROLE_USER])
def confirm_purchase():
    data = request.json
    headers = request.headers
    r = requests.post(product_purchase_api_routes.BASE + product_purchase_api_routes.API +
                         product_purchase_api_routes.CONFIRM_PURCHASE, json=data, headers=headers)
    resp = jsonify(r.json())
    resp.status_code = r.status_code
    return resp


@token_utils.authorization_required(roles=[role.ROLE_USER])
def send_purchase_confirmation_mail():
    data = request.json
    headers = request.headers
    r = requests.post(product_purchase_api_routes.BASE + product_purchase_api_routes.API +
                         product_purchase_api_routes.SEND_PURCHASE_CONFIRMATION_MAIL, json=data, headers=headers)
    resp = jsonify(r.json())
    resp.status_code = r.status_code
    return resp


@token_utils.authorization_required(roles=[role.ROLE_USER])
def get_product_alert_by_product_id_and_user_id():
    args = request.args.to_dict()
    product_id = args.get("productId")
    headers = request.headers
    r = requests.get(product_purchase_api_routes.BASE + product_purchase_api_routes.API +
                        product_purchase_api_routes.GET_PRODUCT_ALERT_BY_PRODUCT_ID_AND_USER_ID +
                        "?productId={}".format(product_id), headers=headers)
    resp = jsonify(r.json())
    resp.status_code = r.status_code
    return resp


@token_utils.authorization_required(roles=[role.ROLE_USER])
def add_product_alert():
    args = request.args.to_dict()
    product_id = args.get("productId")
    headers = request.headers
    r = requests.post(product_purchase_api_routes.BASE + product_purchase_api_routes.API +
                         product_purchase_api_routes.ADD_PRODUCT_ALERT +
                         "?productId={}".format(product_id), headers=headers)
    resp = jsonify(r.json())
    resp.status_code = r.status_code
    return resp


@token_utils.authorization_required(roles=[role.ROLE_EMPLOYEE])
def notify_product_availability():
    args = request.args.to_dict()
    product_id = args.get("productId")
    headers = request.headers
    r = requests.get(product_purchase_api_routes.BASE + product_purchase_api_routes.API +
                        product_purchase_api_routes.NOTIFY_PRODUCT_AVAILABILITY +
                        "?productId={}".format(product_id), headers=headers)
    resp = jsonify(r.json())
    resp.status_code = r.status_code
    return resp


@token_utils.authorization_required(roles=[role.ROLE_ADMIN])
def confirm_payment():
    data = request.json
    headers = request.headers
    r = requests.put(product_purchase_api_routes.BASE + product_purchase_api_routes.API +
                        product_purchase_api_routes.CONFIRM_PAYMENT, json=data, headers=headers)
    resp = jsonify(r.json())
    resp.status_code = r.status_code
    return resp


@token_utils.authorization_required(roles=[role.ROLE_ADMIN])
def send_purchased_digital_video_games():
    data = request.json
    headers = request.headers
    r = requests.post(product_purchase_api_routes.BASE + product_purchase_api_routes.API +
                         product_purchase_api_routes.SEND_PURCHASED_DIGITAL_VIDEO_GAMES, json=data, headers=headers)
    resp = jsonify(r.json())
    resp.status_code = r.status_code
    return resp


@token_utils.authorization_required(roles=[role.ROLE_USER, role.ROLE_ADMIN])
def get_purchase_history():
    args = request.args.to_dict()
    page = args.get("page")
    page_size = args.get("pageSize")
    user_id = args.get("userId")
    headers = request.headers
    r = requests.get(product_purchase_api_routes.BASE + product_purchase_api_routes.API +
                        product_purchase_api_routes.GET_PURCHASE_HISTORY
                        + "?userId={}&page={}&pageSize={}".format(user_id, page, page_size), headers=headers)
    resp = jsonify(r.json())
    resp.status_code = r.status_code
    return resp


@token_utils.authorization_required(roles=[role.ROLE_USER, role.ROLE_ADMIN])
def get_number_of_records_purchase_history():
    args = request.args.to_dict()
    user_id = args.get("userId")
    headers = request.headers
    r = requests.get(product_purchase_api_routes.BASE + product_purchase_api_routes.API +
                        product_purchase_api_routes.GET_NUMBER_OF_RECORDS_PURCHASE_HISTORY +
                        "?userId={}".format(user_id), headers=headers)
    resp = jsonify(r.json())
    resp.status_code = r.status_code
    return resp
